File: ios/Classes/SecureScreenGuardPlugin.swift
import Flutter
import UIKit

public class SecureScreenGuardPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    private static let methodChannelName = "secure_screen_guard/methods"
    private static let eventChannelName  = "secure_screen_guard/events"

    private var eventSink: FlutterEventSink?
    private var widgetProtectionEnabled = false
    private var mode: String = "balanced"
    private var wasRecording = false
    private var recordingCheckTimer: Timer?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = SecureScreenGuardPlugin()

        let methodChannel = FlutterMethodChannel(
            name: methodChannelName,
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: eventChannelName,
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)

        registrar.addApplicationDelegate(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "enable":
            widgetProtectionEnabled = true
            applyProtectionState()
            result(nil)

        case "disable":
            widgetProtectionEnabled = false
            applyProtectionState()
            result(nil)

        case "setMode":
            guard let args = call.arguments as? [String: Any],
                  let mode = args["mode"] as? String else {
                result(FlutterError(code: "INVALID_ARG", message: "mode required", details: nil))
                return
            }
            handleMode(mode)
            result(nil)

        case "isProtected":
            result(shouldProtect())

        case "isRecording":
            if #available(iOS 11.0, *) {
                result(UIScreen.main.isCaptured)
            } else {
                result(false)
            }

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func onListen(withArguments arguments: Any?,
                         eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        applyProtectionState()
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        stopMonitoring()
        return nil
    }

    public func applicationWillResignActive(_ application: UIApplication) {}

    private func shouldProtect() -> Bool {
        switch mode {
        case "strict":
            return true
        case "off":
            return false
        default:
            return widgetProtectionEnabled
        }
    }

    private func applyProtectionState() {
        if shouldProtect() {
            startMonitoring()
        } else {
            stopMonitoring()
        }
    }

    private func startMonitoring() {
        NotificationCenter.default.removeObserver(
            self,
            name: UIApplication.userDidTakeScreenshotNotification,
            object: nil
        )
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(screenshotTaken),
            name: UIApplication.userDidTakeScreenshotNotification,
            object: nil
        )

        if #available(iOS 11.0, *) {
            recordingCheckTimer?.invalidate()
            recordingCheckTimer = Timer.scheduledTimer(
                withTimeInterval: 1.0,
                repeats: true
            ) { [weak self] _ in
                self?.checkRecordingState()
            }
        }
    }

    private func stopMonitoring() {
        NotificationCenter.default.removeObserver(
            self,
            name: UIApplication.userDidTakeScreenshotNotification,
            object: nil
        )
        recordingCheckTimer?.invalidate()
        recordingCheckTimer = nil
        wasRecording = false
    }

    @objc private func screenshotTaken() {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?("screenshot")
        }
    }

    @available(iOS 11.0, *)
    private func checkRecordingState() {
        let recording = UIScreen.main.isCaptured
        if recording && !wasRecording {
            wasRecording = true
            DispatchQueue.main.async { [weak self] in
                self?.eventSink?("recordingStart")
            }
        } else if !recording && wasRecording {
            wasRecording = false
            DispatchQueue.main.async { [weak self] in
                self?.eventSink?("recordingStop")
            }
        }
    }

    private func handleMode(_ mode: String) {
        self.mode = mode
        applyProtectionState()
    }

    deinit {
        stopMonitoring()
    }
}

File: android/src/main/kotlin/com/securescreenguard/secure_screen_guard/SecureScreenGuardPlugin.kt
package com.securescreenguard.secure_screen_guard

import android.app.Activity
import android.view.WindowManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class SecureScreenGuardPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel

    private var activity: Activity? = null
    private var eventSink: EventChannel.EventSink? = null

    private var widgetProtectionEnabled = false
    private var mode: String = "balanced"

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(
            binding.binaryMessenger,
            "secure_screen_guard/methods"
        )
        methodChannel.setMethodCallHandler(this)

        eventChannel = EventChannel(
            binding.binaryMessenger,
            "secure_screen_guard/events"
        )
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, sink: EventChannel.EventSink?) {
                eventSink = sink
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        applyProtectionState()
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        applyProtectionState()
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "enable" -> {
                widgetProtectionEnabled = true
                applyProtectionState()
                result.success(null)
            }
            "disable" -> {
                widgetProtectionEnabled = false
                applyProtectionState()
                result.success(null)
            }
            "setMode" -> {
                val requestedMode = call.argument<String>("mode") ?: "balanced"
                handleMode(requestedMode)
                result.success(null)
            }
            "isProtected" -> result.success(isCurrentlyProtected())
            "isRecording" -> result.success(false)
            else -> result.notImplemented()
        }
    }

    private fun shouldProtect(): Boolean {
        return when (mode) {
            "strict" -> true
            "off" -> false
            else -> widgetProtectionEnabled
        }
    }

    private fun isCurrentlyProtected(): Boolean {
        val window = activity?.window ?: return false
        val flags = window.attributes.flags
        return (flags and WindowManager.LayoutParams.FLAG_SECURE) != 0
    }

    private fun applyProtectionState() {
        if (shouldProtect()) {
            applyFlag()
        } else {
            clearFlag()
        }
    }

    private fun applyFlag() {
        activity?.runOnUiThread {
            activity?.window?.setFlags(
                WindowManager.LayoutParams.FLAG_SECURE,
                WindowManager.LayoutParams.FLAG_SECURE
            )
        }
    }

    private fun clearFlag() {
        activity?.runOnUiThread {
            activity?.window?.clearFlags(WindowManager.LayoutParams.FLAG_SECURE)
        }
    }

    private fun handleMode(mode: String) {
        this.mode = mode
        applyProtectionState()
    }

    private fun notifyEvent(name: String) {
        activity?.runOnUiThread {
            eventSink?.success(name)
        }
    }
}
